das/avail/avail.go

```diff
@@ -0,0 +1,271 @@
+
+outer:
+	for {
+		select {
+		case status := <-sub.Chan():
+			if status.IsInBlock {
+				log.Info("📥  Submit data extrinsic included in block", "blockHash", status.AsInBlock.Hex())
+			}
+			if status.IsFinalized {
+				finalizedblockHash = status.AsFinalized
+				break outer
+			} else if status.IsDropped {
+				return nil, fmt.Errorf("❌ Extrinsic dropped")
+			} else if status.IsUsurped {
+				return nil, fmt.Errorf("❌ Extrinsic usurped")
+			} else if status.IsRetracted {
+				return nil, fmt.Errorf("❌ Extrinsic retracted")
+			} else if status.IsInvalid {
+				return nil, fmt.Errorf("❌ Extrinsic invalid")
+			}
+		case <-timeout:
+			return nil, fmt.Errorf("⌛️  Timeout of %d seconds reached without getting finalized status for extrinsic", a.timeout)
+		}
+	}
+
+	header, err := a.api.RPC.Chain.GetHeader(finalizedblockHash)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get header for finalized block:%+v", err)
+	}
+
+	err = a.vectorx.SubscribeForHeaderUpdate(int(header.Number), 7200)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get the event for header update on vectorx:%+v", err)
+	}
+
+	extrinsicIndex, err := GetExtrinsicIndex(a.api, finalizedblockHash, a.keyringPair.Address, o.Nonce)
+	if err != nil {
+		return nil, err
+	}
+	log.Info("Finalized extrinsic", "extrinsicIndex", extrinsicIndex)
+
+	merkleProofInput, err := QueryMerkleProofInput(finalizedblockHash.Hex(), extrinsicIndex, 1200)
+	if err != nil {
+		return nil, err
+	}
+
+	// Creating BlobPointer to submit over settlement layer
+	blobPointer := BlobPointer{BlockHash: finalizedblockHash, Sender: a.keyringPair.Address, Nonce: nonce, DasTreeRootHash: dastree.Hash(message), MerkleProofInput: merkleProofInput}
+	log.Info("✅  Sucesfully included in block data to Avail", "BlobPointer:", blobPointer)
+	blobPointerData, err := blobPointer.MarshalToBinary()
+	if err != nil {
+		log.Warn("⚠️ BlobPointer MashalBinary error", "err", err)
+		return nil, err
+	}
+
+	return blobPointerData, nil
+}
+
+func (a *AvailDA) Read(ctx context.Context, blobPointer BlobPointer) ([]byte, error) {
+	log.Info("ℹ️ Requesting data from Avail", "BlobPointer", blobPointer)
+
+	// Intitializing variables
+	BlockHash := blobPointer.BlockHash
+	Address := blobPointer.Sender
+	Nonce := gsrpc_types.NewUCompactFromUInt(uint64(blobPointer.Nonce))
+
+	// Fetching block based on block hash
+	avail_blk, err := a.api.RPC.Chain.GetBlock(BlockHash)
+	if err != nil {
+		return []byte{}, fmt.Errorf("❌ cannot get block for hash:%v and getting error:%w", BlockHash.Hex(), err)
+	}
+
+	// Extracting the required extrinsic according to the reference
+	for _, ext := range avail_blk.Block.Extrinsics {
+		// Extracting sender address for extrinsic
+		ext_Addr, err := subkey.SS58Address(ext.Signature.Signer.AsID.ToBytes(), 42)
+		if err != nil {
+			log.Error("❌ unable to get sender address from extrinsic", "err", err)
+		}
+
+		if ext_Addr == Address && ext.Signature.Nonce.Int64() == Nonce.Int64() {
```

Comment:
best to abstract this part into a smaller functions

Reply:
done in commit 154e91e6633633c9ab4fa3227dcfb293e6c8e37f